src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LoadedProject.cs

```diff
@@ -247,29 +247,29 @@ public void Dispose()
         void UpdateProjectSystemProjectCollection<T>(IEnumerable<T> loadedCollection, IEnumerable<T>? oldLoadedCollection, IEqualityComparer<T> comparer, Action<T> addItem, Action<T> removeItem, string logMessage)
         {
             var newItems = new HashSet<T>(loadedCollection, comparer);
-            var oldItems = new HashSet<T>(comparer);
-            var oldItemsCount = oldItems.Count;
+            var oldItems = new HashSet<T>(oldLoadedCollection ?? [], comparer);
 
-            if (oldLoadedCollection != null)
-            {
-                foreach (var item in oldLoadedCollection)
-                    oldItems.Add(item);
-            }
+            var addedCount = 0;
+            var removedCount = 0;
 
             foreach (var newItem in newItems)
             {
                 // If oldItems already has this, we don't need to add it again. We'll remove it, and what is left in oldItems is stuff to remove
                 if (!oldItems.Remove(newItem))
+                {
                     addItem(newItem);
+                    addedCount++;
+                }
             }
 
             foreach (var oldItem in oldItems)
             {
                 removeItem(oldItem);
+                removedCount++;
             }
 
-            if (newItems.Count != oldItemsCount)
-                logger.LogTrace(logMessage, projectFullPathWithTargetFramework, newItems.Count);
+            if (addedCount != 0 || removedCount != 0)
+                logger.LogTrace(logMessage, projectFullPathWithTargetFramework, newItems.Count, addedCount, removedCount);
```

Comment:
Are we already passing a `logMessage` with 4 placeholders? May need to update the comment on line 246.